eZ/Publish/Core/FieldType/Keyword/KeywordStorage/Gateway/LegacyStorage.php

```diff
@@ -286,12 +287,11 @@ protected function insertKeywords(array $keywordsToInsert, $contentTypeId)
         return $keywordIdMap;
     }
 
-    protected function deleteOldKeywordAssignments($fieldId, VersionInfo $versionInfo = null)
+    protected function deleteOldKeywordAssignments($fieldId, VersionInfo $versionInfo)
     {
         $dbHandler = $this->getConnection();
         // do not delete assignments if there's other published version containing this field
-        if ($versionInfo !== null &&
-            $versionInfo->status !== VersionInfo::STATUS_PUBLISHED &&
+        if ($versionInfo->status === VersionInfo::STATUS_ARCHIVED &&
```

Comment:
> Did you mean simplify this line or some db query?

unsure what you mean. I ment you can change to what you have now, the query not beeing needed anymore.

> Done like that, please note that for this to work I had to change checking for archived version status instead of not published

Makes sense
But then update comment to say something like
`// If current version being asked to be deleted is archive, then don't do that if there is some other version which is published (as keyword table is not versioned)`

But it does bring the question how this behaves when deleting a draft, does it delete keywords of the published content?